presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java

```diff
@@ -78,11 +77,11 @@ public Result apply(LimitNode parent, Captures captures, Context context)
         PlanNode left = joinNode.getLeft();
         PlanNode right = joinNode.getRight();
 
-        if (isLeftOrFullOuter(joinNode.getType()) && !isLimited(left, context.getLookup(), parent.getCount())) {
+        if (isLeftOuter(joinNode.getType()) && !isLimited(left, context.getLookup(), parent.getCount())) {
             left = new LimitNode(context.getIdAllocator().getNextId(), left, parent.getCount(), true);
         }
 
-        if (isRightOrFullOuter(joinNode.getType()) && !isLimited(right, context.getLookup(), parent.getCount())) {
+        if (isRightOuter(joinNode.getType()) && !isLimited(right, context.getLookup(), parent.getCount())) {
```

Comment:
Potentially it may result into a performance regression. If the key selected on the right doesn't match any row on the left - the entire right side will have to be streamed before Join outputs any rows. I don't feel strong about it, but I would try to be safe, and add a session property that can be used to disable this optimization in case of emergency.